### PR TITLE
nautilus: mgr/dashboard: Searchable objects for table

### PR DESCRIPTION
**Pull request by shyukri**
backport tracker: https://tracker.ceph.com/issues/42673

---

backport of https://github.com/ceph/ceph/pull/31199
parent tracker: https://tracker.ceph.com/issues/42480

this backport was staged using ceph-backport.sh version 15.0.0.6950
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Comment by shyukri** [comment]
This required some changes until it got green for local tests via ```npm run test```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts [hunk 1]
```diff
@@ -111,85 +103,122 @@ describe('TableComponent', () => {
...
+    it('should find a particular number', () => {
+      doSearch('5', 37, { a: 5, b: 25, c: [-5, "score11"], d: false });
+      // { a: 3, b: 9, c: [ -3, 'score9' ], d: false },
+      // { a: 7, b: 49, c: [ -7, 'score13' ], d: false },
+      // { a: 9, b: 81, c: [ -9, 'score15' ], d: false },
 
-      it('should search for multiple values', () => {
-        doSearch('7 5 3', 5, { a: 57, b: 3249, c: [-7, 'score15'], d: false });
-      });
+      doSearch('9', 50, { a: 3, b: 9, c: [-3, "score9"], d: false });
```
Comment (shyukri) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts]: i'm not sure if that's the expected behavior for the table above

**Comment by smithfarm** [comment]
The make check error appears to be:

```
Linting "ceph-dashboard"...
no-use-before-declare is deprecated. Since TypeScript 2.9. Please use the built-in compiler checks instead.

ERROR: /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts:146:48 - " should be '
ERROR: /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts:151:47 - " should be '

Lint errors found in the listed files.
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts [hunk 2]
```diff
@@ -111,85 +103,122 @@ describe('TableComponent', () => {
...
-          component.updateFilter();
-        };
-        searchTest('a b c', ['a', 'b', 'c']);
-        searchTest('a+b c', ['a+b', 'c']);
-        searchTest('a,,,, b,,,     c', ['a', 'b', 'c']);
-        searchTest('a,,,+++b,,,     c', ['a+++b', 'c']);
-        searchTest('"a b c"   "d e  f", "g, h i"', ['a+b+c', 'd+e++f', 'g+h+i']);
-      });
+    it('should find a particular number', () => {
+      doSearch('5', 37, { a: 5, b: 25, c: [-5, "score11"], d: false });
```
Comment (smithfarm) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts]: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts:146:48 - " should be '

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts (hunk 1 above)
Comment (smithfarm) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts]: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts:151:47 - " should be '

**Comment by smithfarm** [comment]
> This required some changes until it got green for local tests via `npm run test`

It would be nice if you included a description of the changes in `Conflicts` section of the commit message, so reviewers know what you did. For details see https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst#conflict-resolution

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts (hunk 1 above)
Comment (shyukri) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts]: should be fixed

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts (hunk 2 above)
Comment (shyukri) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts]: should be fixed

**Comment by shyukri** [comment]
Regarding linter errors i found https://stackoverflow.com/questions/51621162/typescript-linter-warning-no-unused-variable-is-deprecated-but-im-not-using-t ```but the whole TSLint has been deprecated in favor of typescript-eslint.``` maybe a separate PR for migrating the linter can be added later.

**Comment by tspmelo** [comment]
There is no need to try to fix the linting warning, `no-use-before-declare`, since it doesn't affect the code and linting results.
There has been some discussions (in tslint repo) regarding removing the deprecation of that rule, since the tslint rule works better then the one from typescript compiler.
And we haven't removed that from master.

The changes introduced in a2bb6fccb5dd9e9f7300ed0d25d60df4c7fb2693 should be removed, since they are not related and  don't interfere with the code being backported.

Regarding the migration to ESlint, I have been waiting for Angular to do the migration to start working on that.
Currently there are still from performance issues in typescript-eslint and not all rules are currently supported.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts [hunk 3]
```diff
@@ -111,85 +103,122 @@ describe('TableComponent', () => {
...
+      expect(component.rows[0].d).toBe(false);
+      expect(component.rows[1].d).toBe(false);
+    });
 
-      it('should search if column name is incomplete', () => {
-        doSearch(`'poker array'`, 0);
-        doSearch('pok', 0);
-        doSearch('pok:', 100);
+    it('should test search manipulation', () => {
+      let searchTerms = [];
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts]: Maybe you should have backported another commit before mine?
(The suite works but I'm not sure if that is wanted @smithfarm )

**Comment by yuriw** [comment]
[wip-yuri7-testing-2020-03-19-1610-nautilus](https://trello.com/c/LrERphRb/847-wip-yuri7-testing-2020-03-19-1610-nautilus)

**Pull request merged**